### ISSUE TITLE
Request for Code or Clarification on Reward Recovery (Appendix G.10)

### ISSUE THREAD

**Issue by chufanchen** [hari-sikchi/ReCOIL#3]
Dear authors,

Thank you very much for sharing the codebase for ReCOIL. I’ve been exploring your excellent DualRL paper and am particularly interested in the reward recovery experiment described in **Appendix G.10**.

I’ve been attempting to reproduce the results from that section—specifically, the use of ReCOIL to recover reward signals—but I wasn’t able to locate the corresponding code or scripts for this part in the current repository. Based on the paper’s description, I implemented a version of the reward recovery procedure, which I’ve included below:

```python
import os
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import spearmanr, pearsonr
from absl import app, flags
from itertools import count
import gym
import jax

import wrappers
from dataset_utils import D4RLDataset, split_into_trajectories
from evaluation import evaluate
from learner_imitate import Learner, compute_rewards
from common import Batch
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.atari_wrappers import AtariWrapper
from wrappers.atari_wrapper import (
    ScaledFloatFrame,
    FrameStack,
    FrameStackEager,
    PyTorchFrame,
)
from wrappers.normalize_action_wrapper import check_and_normalize_box_actions
import pickle

FLAGS = flags.FLAGS

flags.DEFINE_string("env_name", "hopper-medium-v2", "Environment name.")
flags.DEFINE_string("model_path", "./tmp/", "Path to the trained model.")
flags.DEFINE_string("save_dir", "./results/correlation/", "Directory to save plots.")
flags.DEFINE_integer("seed", 42, "Random seed.")
flags.DEFINE_integer(
    "eval_episodes", 100, "Number of episodes for correlation analysis."
)
flags.DEFINE_float("gamma", 0.99, "Discount factor.")
flags.DEFINE_float("temp", 1.0, "Loss temperature")
flags.DEFINE_boolean("double", True, "Use double q-learning")


def make_atari(env):
    env = AtariWrapper(env)
    env = PyTorchFrame(env)
    env = FrameStack(env, 4)
    return env


def is_atari(env_name):
    return env_name in [
        "PongNoFrameskip-v4",
        "BreakoutNoFrameskip-v4",
        "SpaceInvadersNoFrameskip-v4",
        "BeamRiderNoFrameskip-v4",
        "QbertNoFrameskip-v4",
        "SeaquestNoFrameskip-v4",
    ]


def make_env(env_name, mointor=True):
    env = gym.make(env_name)
    env = Monitor(env, "gym")
    if is_atari(env_name):
        env = make_atari(env)
    # Normalize box actions to [-1, 1]
    env = check_and_normalize_box_actions(env)
    return env


def make_agent(env, model_path):
    """Load trained agent."""
    observation_sample = env.observation_space.sample()[np.newaxis]
    action_sample = env.action_space.sample()[np.newaxis]

    agent = Learner(
        FLAGS.seed, observation_sample, action_sample, max_steps=100, double_q=FLAGS.double
    )

    agent.load(model_path)
    return agent


def eps(rewards):
    """Calculate episode returns."""
    return [sum(x) for x in rewards]


def part_eps(rewards):
    """Calculate partial episode returns."""
    return [np.cumsum(x) for x in rewards]


def measure_correlations(agent, env, gamma=0.99, episodes=100, save_dir=None):
    """Measure correlations between environment rewards and learned rewards."""
    if save_dir and not os.path.exists(save_dir):
        os.makedirs(save_dir)

    env_rewards = []
    learnt_rewards = []

    for episode in range(episodes):
        part_env_rewards = []
        part_learnt_rewards = []

        state = env.reset()
        done = False
        episode_reward = 0
        episode_learnt_reward = 0

        states = []
        actions = []
        rewards = []
        next_states = []
        masks = []
        dones = []

        # Collect trajectory
        while not done:
            action = agent.sample_actions(
                state,
                temperature=0.0
            )

            next_state, reward, done, _ = env.step(action)

            states.append(state)
            actions.append(action)
            rewards.append(reward)
            next_states.append(next_state)
            masks.append(1.0 - float(done))
            dones.append(float(done))

            q = agent.critic.apply(
                {"params": agent.critic.params}, np.array(state), np.array(action)
            )
            next_v = agent.value.apply(
                {"params": agent.value.params}, np.array(next_state)
            )
            
            y = (1 - done) * gamma * next_v
            irl_reward = jax.numpy.minimum(q[0], q[1]) - y
            # irl_reward = - jax.numpy.minimum(q[0], q[1]) + y

            episode_reward += reward
            part_env_rewards.append(reward)
            part_learnt_rewards.append(irl_reward)
            episode_learnt_reward += irl_reward

            state = next_state

        print(
            f"Episode {episode+1}\tEnvironment rewards: {episode_reward:.2f}\tLearned rewards: {episode_learnt_reward:.2f}"
        )

        env_rewards.append(part_env_rewards)
        learnt_rewards.append(part_learnt_rewards)

    # Calculate correlations
    spearman_corr, spearman_p = spearmanr(eps(learnt_rewards), eps(env_rewards))
    pearson_corr, pearson_p = pearsonr(eps(learnt_rewards), eps(env_rewards))

    print(f"Spearman correlation: {spearman_corr:.4f} (p-value: {spearman_p:.4e})")
    print(f"Pearson correlation: {pearson_corr:.4f} (p-value: {pearson_p:.4e})")

    if save_dir:
        # Plot 1: Episode rewards
        sns.set()
        plt.figure(dpi=150)
        plt.scatter(eps(env_rewards), eps(learnt_rewards), s=10, alpha=0.8)
        plt.xlabel("Environment Returns")
        plt.ylabel("Learned Returns")
        plt.title(f"Episode rewards\nPearson Correlation: {pearson_corr:.4f}")
        plt.savefig(os.path.join(save_dir, "Episode_rewards.png"))
        plt.close()

        # Plot 2: Partial rewards (scatter)
        sns.set()
        plt.figure(dpi=150)
        for i in range(min(20, len(env_rewards))):
            plt.scatter(
                part_eps(env_rewards)[i], part_eps(learnt_rewards)[i], s=5, alpha=0.6
            )
        plt.xlabel("Environment rewards")
        plt.ylabel("Learned rewards")
        plt.savefig(os.path.join(save_dir, "Partial_rewards.png"))
        plt.close()

        # Plot 3: Partial rewards (interpolated)
        sns.set()
        plt.figure(dpi=150)
        for i in range(min(20, len(env_rewards))):
            plt.plot(
                part_eps(env_rewards)[i],
                part_eps(learnt_rewards)[i],
                markersize=1,
                alpha=0.8,
            )
        plt.xlabel("Environment rewards")
        plt.ylabel("Learned rewards")
        plt.savefig(os.path.join(save_dir, "Partial_rewards_interpolated.png"))
        plt.close()

        # Plot 4: Step rewards
        sns.set()
        plt.figure(dpi=150)
        for i in range(min(5, len(env_rewards))):
            plt.scatter(env_rewards[i], learnt_rewards[i], s=5, alpha=0.5)
        plt.xlabel("Environment rewards")
        plt.ylabel("Learned rewards")
        plt.savefig(os.path.join(save_dir, "Step_rewards.png"))
        plt.close()

    return {
        "spearman_corr": spearman_corr,
        "pearson_corr": pearson_corr,
        "env_rewards": env_rewards,
        "learnt_rewards": learnt_rewards,
    }

def normalize(x):
    x = np.array(x)
    return (x - x.mean()) / (x.std() + 1e-8)

def recover_reward(agent, state, action, next_state, done):
    GAMMA = FLAGS.gamma
    q = agent.critic.apply(
        {"params": agent.critic.params}, np.array(state), np.array(action)
    )
    next_v = agent.value.apply(
        {"params": agent.value.params}, np.array(next_state)
    )
    y = (1 - done) * GAMMA * next_v
    irl_reward = jax.numpy.minimum(q[0], q[1]) - y
    return irl_reward

def process_dataset_with_learned_rewards(agent, dataset, save_dir):
    GT_rewards, learnt_rewards = [], []
    for path in dataset:
        traj_len = path["observations"].shape[0]
        path["GT_rewards"] = path["rewards"].copy()
        states = path["observations"][:traj_len]
        actions = path["actions"][:traj_len]
        dones = path["terminals"][:traj_len]
        next_states = np.empty_like(states)
        if traj_len > 1:
            next_states[:-1] = states[1:]
        next_states[-1] = states[-1]
        irl_rewards = recover_reward(agent, states, actions, next_states, dones)
        path["rewards"][:traj_len] = irl_rewards
        GT_rewards.append(path["GT_rewards"].copy())
        learnt_rewards.append(irl_rewards)
        print(f'Path len: {traj_len}, GT rewards: {sum(path["GT_rewards"])}, Recovered rewards: {sum(irl_rewards)}')
    print(f'Pearson correlation: {pearsonr(normalize(eps(learnt_rewards)), normalize(eps(GT_rewards)))}')
    with open(f"D4RL/{FLAGS.env_name}-proxy.pkl", "wb") as f:
        pickle.dump(dataset, f)


def main(_):
    env = make_env(FLAGS.env_name)
    agent = make_agent(env, FLAGS.model_path)
    if not os.path.exists(FLAGS.save_dir):
        os.makedirs(FLAGS.save_dir)
    # Online correlation measurement
    print("Measuring online correlations...")
    online_save_dir = os.path.join(FLAGS.save_dir, "online")
    online_results = measure_correlations(
        agent=agent,
        env=env,
        gamma=FLAGS.gamma,
        episodes=FLAGS.eval_episodes,
        save_dir=online_save_dir,
    )

    # Offline correlation measurement using the dataset
    print("\nAnalyzing offline dataset correlations...")
    dataset_path = f"D4RL/{FLAGS.env_name}.pkl"
    with open(os.path.join(dataset_path), 'rb') as f:
        dataset = pickle.load(f)
    offline_save_dir = os.path.join(FLAGS.save_dir, 'offline')
    process_dataset_with_learned_rewards(
        agent=agent,
        dataset=dataset,
        save_dir=offline_save_dir
    )

    print("\nAnalysis complete. Results saved to:", FLAGS.save_dir)


if __name__ == "__main__":
    app.run(main)

```

Despite multiple runs, I haven’t been able to achieve a correlation close to what was reported in the appendix. I would be sincerely grateful if you could:

1. Let me know if the code for Appendix G.10 is available or planned for release,
2. Share any additional implementation details or experimental configurations that might be relevant,
3. Or provide feedback on whether there might be issues with my current approach.

Thank you again for your contributions to this field—ReCOIL is a fascinating framework, and I greatly appreciate your time and insights.

Best regards,
Chufan Chen
